internal/femto/femto.go

```diff
@@ -3,124 +3,145 @@ package femto
 
 import (
 	"encoding/json"
-	"errors"
 	"log/slog"
 	"net/http"
 	"sync/atomic"
+
+	"github.com/gpuctl/gpuctl/internal/types"
 )
 
 type Femto struct {
 	mux   http.ServeMux
 	reqNo atomic.Uint64
 }
 
-func (f *Femto) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	f.mux.ServeHTTP(w, r)
+type HTTPResponseContent[T any] struct {
+	Headers map[string]string
+	Body    T
+	Status  int
 }
 
-func OnPost[T any, R any](f *Femto, pattern string, handle PostFuncPure[T, R]) {
-	f.mux.HandleFunc(pattern, func(w http.ResponseWriter, r *http.Request) {
-		doPost(f, w, r, handle)
+type PostFunc[T any] func(T, *http.Request, *slog.Logger) (*HTTPResponseContent[types.Unit], error)
+type GetFunc[T any] func(*http.Request, *slog.Logger) (*HTTPResponseContent[T], error)
+
+func (femto *Femto) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	femto.mux.ServeHTTP(w, r)
+}
+
+func OnPost[T any](femto *Femto, pattern string, handle PostFunc[T]) {
+	femto.mux.HandleFunc(pattern, func(writer http.ResponseWriter, request *http.Request) {
+		doPost(femto, writer, request, handle)
 	})
 }
 
-func OnGet[T any](f *Femto, pattern string, handle GetFunc[T]) {
-	f.mux.HandleFunc(pattern, func(w http.ResponseWriter, r *http.Request) {
-		doGet(f, w, r, handle)
+func OnGet[T any](femto *Femto, pattern string, handle GetFunc[T]) {
+	femto.mux.HandleFunc(pattern, func(writer http.ResponseWriter, request *http.Request) {
+		doGet(femto, writer, request, handle)
 	})
 }
 
-func doGet[T any](f *Femto, w http.ResponseWriter, r *http.Request, handle GetFunc[T]) {
-	reqNo := f.nextReqNo()
-	log := f.logger().With(slog.Uint64("req_no", reqNo))
+func generateErrorLogger(log *slog.Logger, writer http.ResponseWriter) func(ctx string, status int, e error) {
+	return func(ctx string, status int, e error) {
+		log.Error(ctx, "err", e)
+		http.Error(writer, e.Error(), status)
+	}
+}
+
+func doGet[T any](femto *Femto, writer http.ResponseWriter, request *http.Request, handle GetFunc[T]) {
```

Comment:
DITTO wrt renames. Makes it hard to see whats changed due to renaming vs others at a glance.